### DISCUSSION TITLE
customizing the Gnode UI

### DISCUSSION THREAD

**Discussion by Adem-Kri** [eclipse-glsp/glsp#992] [Q&A] [answered]
Hello, I'm currently working on elements UI. Among these elements, I'm customizing the Gnode to include a title, which will be represented as a `label`, an `icon` on the left, and a `dropdown` menu in the middle. 

the shape is similar to this:
 ________________________
|           title                  |
| ----------------------- |
|      icon .....  dropdown |
 ----------------------- 

I've looked in the documentation before asking but got no clue concerning elements like radio buttons, dropdowns, etc.

I would appreciate any suggestions.

**Comment by planger** [comment, marked as answer]
Hi @Adem-Kri,

there are two alternatives for implementing HTML inputs, such as drop-downs.
1. [UI Extensions](https://www.eclipse.org/glsp/documentation/ui-extensions/#register-and-activate-custom-ui-extensions)
2. [Foreign elements](https://www.eclipse.org/glsp/documentation/rendering/#foreignobjectviewhttpsgithubcomeclipsesprottyblobmasterpackagessprottysrclibgeneric-viewstsx)


Option 1 is basically a HTML `div` that is rendered on top of diagram elements. It is for instance used for label editing. This approach is very flexible and there is rather convenient API support for it (see [documentation](https://www.eclipse.org/glsp/documentation/ui-extensions/#register-and-activate-custom-ui-extensions)). However, if you want those HTML elements to feel like a part of your diagram element, they have disadvantages, as you'd need to move the `div` alogn the diagram element whenever the diagram element is moved.

With option 2 you'd directly put HTML elements into your SVG code within a [`foreignObject`](https://developer.mozilla.org/en-US/docs/Web/SVG/Element/foreignObject) (i.e. foreign to the SVG namespace). This has the advantage that it is really part of the element and will be moved with the containing diagram element automatically. An example is available in Sprotty with the [SVG example](https://github.com/eclipse-sprotty/sprotty/tree/master/examples/svg), see in particular the [SModel](https://github.com/eclipse-sprotty/sprotty/blob/master/examples/svg/src/standalone.ts) of this example. We use this approach e.g. for line breaks in [multi-line text](https://github.com/eclipse-sprotty/sprotty/pull/171).
You should also be able to directly render the same markup code (`foreignObject` with a HTML namespace) from a view implementation. This way you can also hook up event listeners.
